File: python/spec/queues/queue_list_spec.py
from ivoire import describe, context
from expects import expect
from src.queues.queue_list import QueueList
from src.queues.queue_interface import QueueInterface

with describe(QueueList) as it:
  @it.before
  def before(test):
    test.queue = QueueList()

  with it('inherits from the QueueInterface') as test:
    expect(test.queue).to.be.a(QueueInterface)

  with context(QueueList.__init__):
    with it('creates an empty queue') as test:
      expect(test.queue.list).to.be.empty

  with context(QueueList.isEmpty):
    with it('returns true when queue is empty') as test:
      expect(test.queue.isEmpty()).to.be.true

    with it('returns false when queue is not empty') as test:
      test.queue.insert(1)
      expect(test.queue.isEmpty()).to.be.false

  with context(QueueList.insert):
    with it('adds a new node at the end of the queue') as test:
      test.queue.insert(2)
      test.queue.insert(3)
      expect(test.queue.list[-1].cargo).to.equal(3)

  with context(QueueList.remove):
    with it('removes and returns the node at the top of the queue') as test:
      test.queue.insert(4)
      test.queue.insert(5)
      expect(test.queue.remove().cargo).to.equal(4)

File: ruby/spec/queues/queue_list_spec.rb
require "spec_helper"
require "queues/queue_list"
require "queues/queue_interface"

describe QueueList do
    subject(:queue) { described_class.new }

    it "inherits from QueueInterface class" do
        expect(queue).to be_a QueueInterface
    end

    describe "#initialize" do

        it "initializes an empty queue" do
            expect(queue.list).to be_empty
        end

    end

    describe "#empty?" do

        context "when the queue is empty" do
            it "returns true" do
                expect(queue.empty?).to be_truthy
            end
        end

        context "when the queue is not empty" do
            before do
                queue.insert(22)
            end

            it "returns false" do
                expect(queue.empty?).to be_falsy
            end
        end

    end

    describe "#insert" do
        let(:cargo) { 22 }
        let(:stubbed_node) { Node.new(0) }

        before do
            allow(Node).to receive(:new).and_return(stubbed_node)
        end

        after do
            queue.insert(cargo)
        end

        it "creates a new node from the cargo" do
            expect(Node).to receive(:new).with(cargo)
        end

        it "adds the node to the queue" do
            expect(queue.list).to receive(:push).with(stubbed_node)
        end
    end

    describe "#remove" do
        before { queue.insert(44) }
        after { queue.remove }

        it "removes the top node from the queue" do
            expect(queue.list).to receive(:shift)
        end
    end
end